### DISCUSSION TITLE
Serverless function doesn't work on Netlify

### DISCUSSION THREAD

**Discussion by mrmartineau** [11ty/eleventy#2785] [Q&A] [answered]
Although originally raised in [this issue](https://github.com/11ty/eleventy/issues/2782) as part of another bug I found with Eleventy, I am still in need of assistance. 

I am not sure if my issue is 11ty or Netlify related.

Here's the serverless function and async filter in my [eleventy.js](https://github.com/mrmartineau/11ty-notes/blob/main/.eleventy.js#L38-L51). FYI the search index uses Algolia's [client SDK](https://github.com/algolia/algoliasearch-client-javascript).

```js
eleventyConfig.addPlugin(EleventyServerlessBundlerPlugin, {
  name: 'search',
  functionsDir: './netlify/functions/',
})
eleventyConfig.addAsyncFilter('getResults', function (query) {
  const results = index
    .search(query, {
      attributesToRetrieve: ['title', 'url', 'date', 'tags', 'emoji'],
    })
    .then((res) => {
      return res.hits
    })
  return results
})
```

I followed both [the official docs](https://www.11ty.dev/docs/plugins/serverless/) and [this tutorial](https://www.raymondcamden.com/2020/06/24/adding-algolia-search-to-eleventy-and-netlify) (which is a couple of years old) to setup Algolia search on my site.

The search works perfectly locally, but when deployed to Netlify it does not work. From the Netlify logs, I can see that it times out:

```
Jan 29, 07:57:19 PM: 2376f0de 2023-01-29T19:57:30.939Z 2376f0de-fdf1-467b-a50c-62958f901b64 Task timed out after 10.01 seconds
Jan 29, 07:57:29 PM: 2376f0de Duration: 10013.75 ms	Memory Usage: 104 MB	Init Duration: 835.66 ms
```

I am using Netlify's free tier, I have added the 3 environment variables so I can't think what other setup I need in order to get this to work.

I think the free tier should be able to handle the small use-case that I have.

Is there anything obvious that I might be missing?

- Repo: https://github.com/mrmartineau/11ty-notes
- eleventy.js: https://github.com/mrmartineau/11ty-notes/blob/main/.eleventy.js
- search template that uses the async filter: https://github.com/mrmartineau/11ty-notes/blob/main/src/search.njk

**Comment by zachleat** [comment]
I wonder if @cfjedimaster might be able to pitch in here, as their tutorial was mentioned.  My hunch is that it’s probably doing some heavy data fetching that is using up the runtime quota, stuff that might be cached locally?

**Comment by cfjedimaster** [comment]
Search itself should be super fast - to be clear, does the site _build_ ok, and you only get the error trying to perform the search?

**Reply by mrmartineau** [reply]
Yes. the site builds fine (both locally and on Netlify) and the search works perfectly locally. It's only when I try searching on the hosted version that it breaks. Here's an example search page: https://notez.zander.wtf/search/?query=css (should have linked to that originally of course 🤦)

**Reply by cfjedimaster** [reply]
Oh wait - I think ... well... I mean I don't know how it's working locally. Your code isn't awaiting for the result. I think you want this:

```
eleventyConfig.addAsyncFilter('getResults', function (query) {
  const results = await index
    .search(query, {
      attributesToRetrieve: ['title', 'url', 'date', 'tags', 'emoji'],
    })
  return results.hits;
})
```

I'm typing this in here, so forgive the typo, but I think that was the issue. You main need to add async before function(query) too.

**Reply by mrmartineau** [reply]
Converting to an async/await style rather than promises shouldn't change anything, because it is essentially the same code, but I will give it a go.. btw this is what I used, which works locally:

```js
  eleventyConfig.addAsyncFilter('getResults', async function (query) {
    const results = await index.search(query, {
      attributesToRetrieve: ['title', 'url', 'date', 'tags', 'emoji'],
    })
    return results.hits
  })
```

..but still does not work when published  😞

**Reply by cfjedimaster** [reply]
When I look at your repo, I don't see that though. Are you committing your changes to GitHub and is Netlify pulling from that?

**Reply by mrmartineau** [reply]
I realise I never added a `catch` block to see what the error was, so I added one and a bunch of logging, but it still times out. None of the logging shows at all..

```js
eleventyConfig.addAsyncFilter('getResults', function (query) {
  console.log(`🚀 ~ query`, query)
  const results = index
    .search(query, {
      attributesToRetrieve: ['title', 'url', 'date', 'tags', 'emoji'],
    })
    .then((res) => {
      console.log(`🚀 ~ .then ~ res.hits`, res.hits)
      return res.hits
    })
    .catch((err) => {
      console.log(err)
    })
  return results
})
```

Is using a filter the correct thing to do here? Is there another way to achieve what I want?

**Reply by cfjedimaster** [reply]
Your code is still not right though. You need to await the result of index.search, get rid of .then, and return results.hits. See the example I provided earlier.

**Reply by mrmartineau** [reply]
> When I look at your repo, I don't see that though. Are you committing your changes to GitHub and is Netlify pulling from that?

@cfjedimaster I have been [committing fairly quickly](https://github.com/mrmartineau/11ty-notes/commits/main) over the past few minutes to try and debug this issue. Each time it works locally and fails when published. Think I'm going to remove all algolia stuff from the async filter and check that that part is working now..

**Reply by mrmartineau** [reply]
> Your code is still not right though. You need to await the result of index.search, get rid of .then, and return results.hits. See the example I provided earlier.

I tried that [already](https://github.com/mrmartineau/11ty-notes/commit/d2b5185ff5f59d0f289398efae5ab1a8f9ce7879), same result unfortunately.

**Reply by mrmartineau** [reply]
@cfjedimaster btw thanks very much for helping on this.

**Reply by cfjedimaster** [reply]
i'll take a look tomorrow - signing off tech for the night :)

**Reply by cfjedimaster** [reply]
Question - when you run locally, are you using `ntl dev`? I'm not seeing that in your package.json set of scripts.

**Reply by cfjedimaster** [reply]
Ok, I got it working. I honestly didn't know 11ty Serverless worked outside of ntl dev, but that's ok. Here's the code I used:

```
  eleventyConfig.addFilter('getResults', async function (query) {
    console.log(`🚀 ~ query`, query)
    const results = await index
       .search(query, {
         attributesToRetrieve: ['title', 'url', 'date', 'tags', 'emoji'],
       });
      
    return results.hits;
  })
  ```

**Reply by mrmartineau** [reply]
@cfjedimaster I use `npm run dev` locally mostly and have also tried the Netlify CLI, both of which have always worked using the async/await or promise/then methods. I think the issue must be on the Netlify side so have created [a forum post](https://answers.netlify.com/t/function-has-crashed-with-eleventy/84625) for it. When I get to the bottom of this, I will make sure I respond with the solution.

And thank you again for helping with this, truly above and beyond 🙏

**Reply by cfjedimaster** [reply]
So is my change working for you?

**Reply by mrmartineau** [reply]
It works locally - both versions work, but not on Netlify.

**Reply by cfjedimaster** [reply]
I assume you set your environment variables properly, right? If I remember right, Netlify lets you add users to a site, would you feel comfortable adding me so I can take a look? We can switch out to email if you want (raymondcamden@gmail.com), and then update this thread when we figre out the final issue.

**Reply by mrmartineau** [reply]
I have checked my env vars and am fairly certain that they are not the issue. I wouldn't mind giving you access, but I will wait to see if my forum post yields anything first.

**Reply by mrmartineau** [reply]
So I have some updates.. I still have not resolved my issue, but I have got a lot closer.

My conversation with Netlify support gave me an idea. I hadn't realised that the 11ty filter that I was using `getResults` would never display any of the logs that I had in place because they are not run in the context of the serverless function. So I moved all my algolia searching stuff (and a bunch of debug logs) into [the serverless function](https://github.com/mrmartineau/11ty-notes/blob/main/netlify/functions/searcher/index.js) itself.

Even on local, I quickly found that the Shiki Twoslash code highlighting plugin was a problem, so I disabled it and everything started working! But only locally. 

Once deployed to Netlify, with Shiki Twoslash disabled, a new issue presented itself.. and it is not a time out issue; you can see the output below. 

```
Feb 6, 01:28:14 PM: INIT_START Runtime Version: nodejs:16.v11	Runtime Version ARN: arn:aws:lambda:us-east-1::runtime:ee071345fbee0f4751fcbcd47a5784052d8f7d4256ca7f31e6056b294cee515c
Feb 6, 01:28:15 PM: 4aa77591 INFO   🚀 ~ handler ~ searchQuery [ 'css' ]
Feb 6, 01:28:15 PM: 4aa77591 INFO   🚀 ~ handler ~ results [
  {
    title: 'CSS font-stacks',
    url: '/notes/css-font-stacks/',
    date: '2020-02-27T00:00:00.000Z',
    tags: [ 'css' ],
    objectID: '44fff048fada8_dashboard_generated_id'
  },
  {
    title: 'CSS Protips',
    url: '/notes/css-protips/',
    date: '2023-01-21T09:05:03.000Z',
    tags: [ 'css' ],
    objectID: '1f29f2f94e186b_dashboard_generated_id'
  },
]

Feb 6, 01:28:15 PM: 4aa77591 INFO   searcher - algolia: 303.783ms
Feb 6, 01:28:15 PM: 4aa77591 ERROR  [11ty] Problem writing Eleventy templates: (more in DEBUG output)
Feb 6, 01:28:15 PM: 4aa77591 ERROR  [11ty] Cannot read properties of null (reading 'toString') (via TypeError)
Feb 6, 01:28:15 PM: 4aa77591 ERROR  [11ty] 
[11ty] Original error stack trace: TypeError: Cannot read properties of null (reading 'toString')
[11ty]     at getGitLastUpdatedTimeStamp (/var/task/node_modules/@11ty/eleventy/src/Util/DateGitLastUpdated.js:17:16)
[11ty]     at module.exports (/var/task/node_modules/@11ty/eleventy/src/Util/DateGitLastUpdated.js:24:19)
[11ty]     at Template.getMappedDate (/var/task/node_modules/@11ty/eleventy/src/Template.js:941:17)
[11ty]     at Template.addPageDate (/var/task/node_modules/@11ty/eleventy/src/Template.js:403:30)
[11ty]     at Template.getData (/var/task/node_modules/@11ty/eleventy/src/Template.js:390:29)
[11ty]     at async TemplateMap.add (/var/task/node_modules/@11ty/eleventy/src/TemplateMap.js:65:16)
[11ty]     at async TemplateWriter._addToTemplateMap (/var/task/node_modules/@11ty/eleventy/src/TemplateWriter.js:235:7)
[11ty]     at async TemplateWriter._createTemplateMap (/var/task/node_modules/@11ty/eleventy/src/TemplateWriter.js:325:5)
[11ty]     at async TemplateWriter.generateTemplates (/var/task/node_modules/@11ty/eleventy/src/TemplateWriter.js:360:5)
[11ty]     at async TemplateWriter.getJSON (/var/task/node_modules/@11ty/eleventy/src/TemplateWriter.js:418:20)

Feb 6, 01:28:15 PM: 4aa77591 INFO   Serverless Error: TypeError: Cannot read properties of null (reading 'toString')
    at getGitLastUpdatedTimeStamp (/var/task/node_modules/@11ty/eleventy/src/Util/DateGitLastUpdated.js:17:16)
    at module.exports (/var/task/node_modules/@11ty/eleventy/src/Util/DateGitLastUpdated.js:24:19)
    at Template.getMappedDate (/var/task/node_modules/@11ty/eleventy/src/Template.js:941:17)
    at Template.addPageDate (/var/task/node_modules/@11ty/eleventy/src/Template.js:403:30)
    at Template.getData (/var/task/node_modules/@11ty/eleventy/src/Template.js:390:29)
    at async TemplateMap.add (/var/task/node_modules/@11ty/eleventy/src/TemplateMap.js:65:16)
    at async TemplateWriter._addToTemplateMap (/var/task/node_modules/@11ty/eleventy/src/TemplateWriter.js:235:7)
    at async TemplateWriter._createTemplateMap (/var/task/node_modules/@11ty/eleventy/src/TemplateWriter.js:325:5)
    at async TemplateWriter.generateTemplates (/var/task/node_modules/@11ty/eleventy/src/TemplateWriter.js:360:5)
    at async TemplateWriter.getJSON (/var/task/node_modules/@11ty/eleventy/src/TemplateWriter.js:418:20)
Feb 6, 01:28:15 PM: 4aa77591 Duration: 492.26 ms	Memory Usage: 106 MB	Init Duration: 769.38 ms
```

I use `git Last Modified` for many dates of the files in this repo, but on this page, as far as I know, it should only be rendering dates from the search results.

I even added `date: git Last Modified` to the  front-matter of my search template, but it did not resolve the issue.

```njk
---
layout: layouts/base.njk
title: Search Page
permalink:
  searcher: /search/
date: git Last Modified
---

<div class="container">
  <div class="prose prose-zinc dark:prose-invert mb-4">
    <h2 class="mb-3">
      You searched for: "{{ eleventy.serverless.query.query }}"
    </h2>
  </div>
  {% include 'components/note-list-search.njk' %}
</div>
```

Looking at the [source code](https://github.com/11ty/eleventy/blob/97243b6a112936a4828c67350738e085098d23dd/src/Util/DateGitLastUpdated.js#L7) for the `getGitLastUpdatedTimeStamp` function, it is spawning git to get the date. Why is it doing this on a serverless function?

One bonus is that the calls to Algolia are very quick, so I don't think I'll need to worry about time out issues with this method in the future.

**Reply by cfjedimaster** [reply]
From what I know, you aren't supposed to actually touch the function Eleventy generates. That function's purpose is to 'connect' the serverless call to an Eleventy template. It's basically a way to run an Eleventy template _after_ a build has been done. I'd remove that code and go back to having the filter as you had. I'm truly at a loss as to why this fails in production, but I think you need to undo the changes made to the serverless function.

**Reply by mrmartineau** [reply]
@zachleat would you be able to guide us here? We should be able to do whatever we want in the serverless function, right? That's why it is committed to git and everything else related to 11ty serverless functions are ignored. Very keen to understand this, and keen to see if you have any thoughts on my latest errors..

**Reply by cfjedimaster** [reply]
I did some digging, adding some console msgs to netlify/functions/search/index.js, and this is the line that is timing out:

```
let [page] = await elev.getOutput();
```

I've reduced search.njk to the bare minimum, and it still times out. Very weird.

**Reply by cfjedimaster** [reply]
Further more, I comment that out, and set page.content to poo, and it can be returned just fine.

**Reply by cfjedimaster** [reply]
I've even tried setting DEBUG=Eleventy:Serverless, and honestly I can't see a real difference between my local env and Netlify. I believe you said you also reached out to Netlify, if you used their support forums can you share the url?

**Reply by mrmartineau** [reply]
In a fit of "try anything to fix it", I changed the name of the function to [`searcher`](https://github.com/mrmartineau/11ty-notes/blob/main/netlify/functions/searcher/index.js), so there is no `netlify/functions/search/index.js` anymore (I linked to it above actually).

Given that disabling Shiki Twoslash improved the outcome somewhat, I'm inclined to think that the disabling even more plugins will help matters. Mind you, I am pretty new to 11ty so it is only a guess. 

Do you know if a different 11ty config be used for serverless functions? or can I disable certain parts of the config using the `config` option in my function?

**Reply by cfjedimaster** [reply]
I can say if you are willing to switch to client-side JavaScript, that would get it working rather quickly. Obviously *something* is off here and it should be working, and we may have found a real bug in the plugin, or perhaps on Netlify. But if you are willing to forgo the serverless aspect and just use client-side JS, that would definitely get you up and running today. 

As for the Shiki issue... when I deployed to Netlify, I didn't see that issue. I'm still running my copy of your site from when I first suggested a fix, so maybe I'm more behind than I think.

**Reply by mrmartineau** [reply]
> so maybe I'm more behind than I think.

I think you must be. I've committed loads of random changes in an effort to fix this.

Regarding the client-side suggestion: I moved this site from [Gatsby](https://notes.zander.wtf/) to 11ty to reduce client-side JS as much as possible. I will keep persevering... 🤞

**Reply by mrmartineau** [reply]
Also, regarding Shiki: it works fine when viewing all the pages, but gave errors in the serverless function which led me to remove it as described above.

**Reply by cfjedimaster** [reply]
Oh hey! So I did that, removed the plugin, and now things are working for me. So maybe try the mod I shared above, keep shiki out, and see if it works? Then we can file a bug report.

**Comment by mrmartineau** [comment, marked as answer]
FYI after a lot of debugging I finally figured out what was at the root of this issue. It was related to the `singleTemplateScope` setting on the `EleventyServerless` function. Because I set that value to `false` and used `date: git Last Modified` for a bunch of notes, an internal function was being run on those notes to get the date but it didn't have an git information so returned an error. I'm not certain if this is a bug so raised an issue [here](https://github.com/11ty/eleventy/issues/2796#issuecomment-1424965326) about it.

Thanks again to @cfjedimaster for all your help debugging this.

**Reply by cfjedimaster** [reply]
Thank you for the update!
